src/transform/src/lib.rs

```diff
@@ -318,42 +318,51 @@ impl Transform for Fixpoint {
         // a bug somewhere that prevents the relation from settling on a
         // stable shape.
         let mut iter_no = 0;
+        let mut seen = BTreeSet::new();
+        seen.insert(relation.clone());
         loop {
-            let start_size = relation.size();
+            let prev_size = relation.size();
             for i in iter_no..iter_no + self.limit {
                 let original = relation.clone();
                 self.apply_transforms(relation, ctx, format!("{i:04}"))?;
                 if *relation == original {
                     mz_repr::explain::trace_plan(relation);
                     return Ok(());
                 }
+                if !seen.insert(relation.clone()) {
+                    // We got into an infinite loop (e.g., we are oscillating between two plans).
+                    // This is not catastrophic, because we can just say we are done now,
+                    // but it would be great to eventually find a way to prevent these loops from
+                    // happening in the first place. We have several relevant issues, see
+                    // https://github.com/MaterializeInc/materialize/issues/27954#issuecomment-2200172227
+                    mz_repr::explain::trace_plan(relation);
+                    soft_panic_or_log!(
+                        "Fixpoint {} detected a loop after {i} iterations",
+                        self.name
+                    );
+                    return Ok(());
+                }
             }
-            let final_size = relation.size();
+            let current_size = relation.size();
 
             iter_no += self.limit;
 
-            if final_size < start_size {
+            if current_size < prev_size {
                 tracing::warn!(
-                    "fixpoint {} ran for {} iterations \
+                    "Fixpoint {} ran for {iter_no} iterations \
```

Comment:
Changed all to `{}`. (It's unfortunately not possible to have the `{...}` version in all of these, because the `{...}` doesn't allow non-trivial expressions.)